### PR TITLE
Use vctrs to validate indices and positions

### LINKED ISSUE #119: Formalise object not found error and improve error message
```r
vars_select(letters, foo)
#> Error: object 'foo' not found

vars_select(letters, "foo")
#> Error: Unknown column `foo`

vars_select(letters, one_of("foo"))
#> named character(0)
#> Warning message:
#> Unknown columns: `foo`
```

Out of bounds:

```r
vars_select(letters, 100)
#> Error: Can't select column because the data frame is too small.
#> These indices are too large: 100

vars_select(letters, -100)
#>   a   b   c   d   e   f   g   h   i   j   k   l   m   n   o   p   q   r   s
#> "a" "b" "c" "d" "e" "f" "g" "h" "i" "j" "k" "l" "m" "n" "o" "p" "q" "r" "s"
#>   t   u   v   w   x   y   z
#> "t" "u" "v" "w" "x" "y" "z"
```

### PR DESCRIPTION
**Pull request by lionel-**
Requires r-lib/vctrs#616.
Closes #119.

Following is a selection of vctrs-powered errors. They seem to be consistent and informative. All these errors inherit from `tidyselect_error_index`.

1. `vars_select()`

```r
"Bare names"

vars_select(letters, foo)
#> Error: Must select existing columns.
#> x Can't subset element with unknown name `foo`.

vars_select(letters, -foo)
#> Error: Must select existing columns.
#> x Can't subset element with unknown name `foo`.


"Names"

vars_select(letters, "foo")
#> Error: Must select existing columns.
#> x Can't subset element with unknown name `foo`.

vars_select(letters, a:"foo")
#> Error: Must select existing columns.
#> x Can't subset element with unknown name `foo`.


"Positions"

vars_select(letters, 30, 50, 100)
#> Error: Must select existing columns.
#> x Can't subset positions 30, 50 and 100.
#> i There are only 26 elements.

vars_select(letters, -100)
#> Error: Must select existing columns.
#> x Can't subset position 100.
#> i There are only 26 elements.

vars_select(letters, 2.5)
#> Error: Must select with column names or positions.
#> x Lossy cast from `i` <double> to <integer>.
```

2. `vars_rename()`

```r
vars_rename(letters, A = TRUE)
#> Error: Must select with column names or positions.
#> x `i` has the wrong type `logical`.
#> i Positions and names must be integer or character.

vars_rename(letters, A = 1.5)
#> Error: Must select with column names or positions.
#> x Lossy cast from `i` <double> to <integer>.
```

3. `vars_pull()`

```r
vars_pull(letters, letters)
#> Error: Must select with column names or positions.
#> x `var` has the wrong size 26.
#> i Positions and names must be size 1.

vars_pull(letters, NA)
#> Error: Must select with column names or positions.
#> x `var` can't be `NA`.
#> i Positions and names can't be missing.

vars_pull(letters, "foo")
#> Error: Must select existing columns.
#> x Can't subset element with unknown name `foo`.

vars_pull(letters, -100)
#> Error: Must select existing columns.
#> x Can't subset position 100.
#> i There are only 26 elements.

vars_pull(letters, -Inf)
#> Error: Must select with column names or positions.
#> x Lossy cast from `var` <double> to <integer>.
```

`vars_rename()` and `vars_pull()` currently still throw "object not found" errors rather than OOB errors.

### CODE REVIEW

**Comment by lionel-** [comment]
The `cnd_issue()` method is overridden to mention data frames and columns. The `cnd_bullets()` implementation is left to vctrs.

File: R/conditions.R [hunk 1]
```diff
@@ -0,0 +1,54 @@
+
+subclass_index_errors <- function(expr) {
+  tryCatch(
+    expr,
+    vctrs_error_index_oob = function(cnd) {
```
Comment (hadley) [inline R/conditions.R]: Why not catch `vctrs_error_index_oob_names ` separately?

File: tests/testthat/outputs/vars-pull-input-checking.txt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+> vars_pull(letters, TRUE)
+Error: Must select with column names or positions.
+x `var` has the wrong type `logical`.
+i This index must be a position or a name.
+
+> vars_pull(letters, NA)
+Error: Must select with column names or positions.
+x `var` can't be `NA`.
+i This index can't be missing.
```
Comment (hadley) [inline tests/testthat/outputs/vars-pull-input-checking.txt]: This seems a bit uninformative, but I think it's fine to keep if it's hard to eliminate.

File: R/conditions.R (hunk 1 above)
Comment (lionel-) [reply R/conditions.R]: oh yes that makes more sense.

**Pull request merged**